src/backend/opencl/compile_module.cpp

```diff
@@ -141,10 +202,45 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
 
 Module loadModuleFromDisk(const int device, const string &moduleKey,
                           const bool isJIT) {
-    UNUSED(device);
-    UNUSED(moduleKey);
-    UNUSED(isJIT);
-    return {};
+    const string &cacheDirectory = getCacheDirectory();
+    if (cacheDirectory.empty()) return Module{};
+
+    auto &dev              = opencl::getDevice(device);
+    const string cacheFile = cacheDirectory + AF_PATH_SEPARATOR +
+                             getKernelCacheFilename(device, moduleKey);
+    try {
+        std::ifstream in(cacheFile, std::ios::binary);
+        if (!in.is_open()) {
+            AF_ERROR("Unable to open binary cache file", AF_ERR_INTERNAL);
+        }
+        in.exceptions(std::ios::failbit | std::ios::badbit);
+
+        // TODO Handle cases where program objects are created from contexts
+        // having multiple devices
+        size_t clbinHash = 0;
+        in.read(reinterpret_cast<char *>(&clbinHash), sizeof(clbinHash));
+        size_t clbinSize = 0;
+        in.read(reinterpret_cast<char *>(&clbinSize), sizeof(clbinSize));
+        vector<unsigned char> clbin(clbinSize);
+        in.read(reinterpret_cast<char *>(clbin.data()), clbinSize);
+        in.close();
+
+        const size_t recomputedHash =
+            deterministicHash(clbin.data(), clbinSize);
+        if (recomputedHash != clbinHash) {
+            AF_ERROR("Binary on disk seems to be corrupted", AF_ERR_LOAD_SYM);
+        }
+        Program progWithBin(opencl::getContext(), {dev}, {clbin});
+        progWithBin.build();
+
+        AF_TRACE("{{{:<30} : loaded from {} for {} }}", moduleKey, cacheFile,
+                 dev.getInfo<CL_DEVICE_NAME>());
+        return {progWithBin};
+    } catch (...) {
```

Comment:
Can you be more specific which exceptions you catch. We should avoid doing this in the backend because it masks errors in the code.

Reply:
The only two specific exceptions we can catch here are `ios_base::failure` and `cl::Error`. In both cases, we would end up returning `AF_ERR_INTERNAL` error code unless we expand our existing codes. Isn't it ? Like add `AF_ERR_DISK_IO`. Final catch would have to be `...` to avoid throwing exception from this location because the inability to cache binary to disk is not a blocker for ArrayFire to run.